File: jifrozen/이진탐색/고정점_찾기.java
package 이진탐색;
import java.util.*;
import java.io.*;

public class 고정점_찾기 {
	public static int bs(int[] arr,int start,int end){
		if(start>end) return -1;
		int mid = (start + end) / 2;
		if (arr[mid] == mid)
			return mid;
		else if (arr[mid] < mid)
			return bs(arr, mid + 1, end);
		else
			return bs(arr, start, mid - 1);
	}
	public static void main(String[] args) throws Exception{
		BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
		int n=Integer.parseInt(br.readLine());

		StringTokenizer st=new StringTokenizer(br.readLine());
		int[] arr=new int[n];
		for(int i=0;i<n;i++){
			arr[i]=Integer.parseInt(st.nextToken());
		}
		int result=bs(arr,0,n);

		System.out.print(result);
	}
}

File: yunva17/이진탐색/고정점_찾기.py
import sys
input = sys.stdin.readline

n = int(input())
array = list(map(int, input().split()))

def binary_search(array, start, end):

    while start <=end:
        mid = (start+end)//2

        if mid == array[mid]:
            return mid
        elif mid > array[mid]:
            start = mid+1
        else:
            end = mid-1

    if mid == None:
            return -1
    
result = binary_search(array, 0, n-1)

if  result == None:
    print(-1)
else:
    print(result)
